README.md

```diff
@@ -0,0 +1,150 @@
+# NimbleHex
+
+NimbleHex allows self-hosting of Hex packages.
+
+Features:
+
+* Pluggable storage. NimbleHex ships with local filesystem and S3 implementations
+* Publishing packages via HTTP API
+* Support for multiple repositories
+* Mirroring
+
+## Setup
+
+Clone and install dependencies:
+
+    git clone git@github.com:wojtekmach/nimble_hex.git
+    cd nimble_hex
+    mix deps.get
+
+Start a development server:
+
+    iex -S mix
+
+By default, the development server is configured with two repositories:
+
+1. `test_repo` is a custom repository
+2. `hexpm_mirror` is a mirror of the official Hex.pm repository, configured to only fetch package
+   "decimal".
+
+Both repositories are configured to store files locally. See `config/config.exs` and `config/dev.exs` for more information.
+
+Now, let's create a new package and publish it to our `test_repo` repository:
+
+    mix new foo
+    cd foo
+
+Make following changes to `mix.exs`:
+
+```elixir
+  def project() do
+    [
+      # ...
+      description: "Some description",
+      package: package(),
+      hex: hex()
+    ]
+  end
+
+  defp package() do
+    [
+      licenses: ["Apache-2.0"],
+      links: %{}
+    ]
+  end
+
+  defp hex() do
+    [
+      api_url: "http://localhost:4000/api/repos/test_repo",
+      api_key: "does-not-matter"
+    ]
+  end
+```
+
+Now publish the package:
+
+    mix hex.publish package
+
+Finally, in order to resolve the package, configure Hex to use the custom repository:
+
+    cd /path/to/nimble_hex
+    mix hex.repo add test_repo http://localhost:4000/repos/test_repo --public-key priv/test_repo_public.pem
+
+And add it as a dependency from another Mix project:
+
+```elixir
+  defp deps() do
+    {:foo, "~> 0.1", repo: "test_repo"}
+  end
+```
+
+Since the development server includes a `hexpm_mirror` repo, let's try that too:
+
+    HEX_MIRROR_URL=http://localhost:4000/repos/hexpm_mirror mix hex.package fetch decimal 1.8.0
+
+See [`mix help hex.config`](https://hexdocs.pm/hex/Mix.Tasks.Hex.Config.html) for more information
+about configuring your Hex installation.
+
+## Deploying with releases
```

Comment:
I feel like this section is unnecessarily complicated by the S3 setup, do you think we should split it up?  I wanted to have s3 setup committed to code so that I can easily test it locally, but at the same time I think people may not want to use it, so perhaps it's better to default release setup to use local filesystem adapter instead?

Or maybe we assemble two different releases, one with local storage backend and one with S3? That seems it may overly complicate things too, not sure.